### DISCUSSION TITLE
Way to run an async function immediately?

### DISCUSSION THREAD

**Discussion by Mummified316** [Rapptz/discord.py#7467] [Q&A] [answered]
Is it possible to run an async function such as guild.create_text_channel() immediately rather being put at the back of the queue? Particularly, I want to make sure everything in my on_ready() function finishes running before any other events are processed.

**Comment by SebbyLaw** [comment]
If all you're doing is making HTTP requests, you can just manually make those requests before starting up the bot. See how it's done under the hood in [`http.py`](https://github.com/Rapptz/discord.py/blob/master/discord/http.py#L865).

**Reply by Mummified316** [reply]
Unfortunately, as far as I am aware, I cannot do things such as bot.get_channel() before the bot has started. Either way, I also need to run async functions immediately when certain events are called.

**Comment by SebbyLaw** [comment, marked as answer]
> Unfortunately, as far as I am aware, I cannot do things such as bot.get_channel() before the bot has started. Either way, I also need to run async functions immediately when certain events are called.

I'm writing this in another comment because this is probably the answer you're looking for.

You could create something that acts similar to `bot.wait_until_ready()` using an [`asyncio.Event`](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event).

For example:

```py
import asyncio
import discord

class MyBot(discord.Client):
    def __init__(self):
        super().__init__()
        self.my_event = asyncio.Event()

    async def wait_until_actually_ready(self):
        return await self.my_event.wait()

    async def on_ready(self):
        ... # do whatever processing you need to here
        # self.get_channel() etc etc
        self.my_event.set()

    async def on_message(self, message):
        await self.wait_until_actually_ready()  # wait until we've done things in on_ready
        ... # your actual event listener code

bot = MyBot()
bot.run(...)
```

This will make your `on_message` event listener code only run once you set the Event in on_ready.
